cygwin/CygwinProcessTable.c

```diff
@@ -0,0 +1,799 @@
+#endif
+   if (statok == -1)
+      return false;
+
+   if (process->st_uid != sstat.st_uid) {
+      process->st_uid = sstat.st_uid;
+      process->user = UsersTable_getRef(host->usersTable, sstat.st_uid);
+   }
+
+   return true;
+}
+
+static void CygwinProcessTable_readMaps(CygwinProcess* process, openat_arg_t procFd, bool checkDeletedLib) {
+   Process* proc = (Process*)process;
+
+   proc->usesDeletedLib = false;
+
+   FILE* mapsfile = fopenat(procFd, "maps", "r");
+   if (!mapsfile)
+      return;
+
+   char buffer[1024];
+   while (fgets(buffer, sizeof(buffer), mapsfile)) {
+      bool map_execute;
+      unsigned int map_devmaj;
+      unsigned int map_devmin;
+      uint64_t map_inode;
+
+      // Short circuit test: Look for a slash
+      if (!strchr(buffer, '/'))
+         continue;
+
+      // Parse format: "%Lx-%Lx %4s %x %2x:%2x %Ld"
+      char* readptr = buffer;
+
+      fast_strtoull_hex(&readptr, 16);
+      if ('-' != *readptr++)
+         continue;
+
+      fast_strtoull_hex(&readptr, 16);
+      if (' ' != *readptr++)
+         continue;
+
+      map_execute = (readptr[2] == 'x');
+      readptr += 4;
+      if (' ' != *readptr++)
+         continue;
+
+      while (*readptr > ' ')
+         readptr++; // Skip parsing this hex value
+      if (' ' != *readptr++)
+         continue;
+
+      map_devmaj = fast_strtoull_hex(&readptr, 4);
+      if (':' != *readptr++)
+         continue;
+
+      map_devmin = fast_strtoull_hex(&readptr, 4);
+      if (' ' != *readptr++)
+         continue;
+
+      //Minor shortcut: Once we know there's no file for this region, we skip
+      if (!map_devmaj && !map_devmin)
+         continue;
+
+      map_inode = fast_strtoull_dec(&readptr, 20);
+      if (!map_inode)
+         continue;
+
+      if (checkDeletedLib && map_execute && !proc->usesDeletedLib) {
+         while (*readptr == ' ')
+            readptr++;
+
+         if (*readptr != '/')
+            continue;
+
+         if (strstr(readptr, "$Recycle.Bin")) {
+            proc->usesDeletedLib = true;
+            break;
+         }
+      }
+   }
+
+   fclose(mapsfile);
+
+   // m_lrs is set in CygwinProcessTable_readStatmFile from /proc/statm
+}
+
+static bool CygwinProcessTable_readStatmFile(CygwinProcess* process, openat_arg_t procFd, const CygwinMachine* host) {
+   FILE* statmfile = fopenat(procFd, "statm", "r");
+   if (!statmfile)
+      return false;
+
+   long int dummy;
+
+   int r = fscanf(statmfile, "%ld %ld %ld %ld %ld %ld %ld",
+                  &process->super.m_virt,
+                  &process->super.m_resident,
+                  &process->m_share,
+                  &process->m_trs,
+                  &process->m_lrs,
+                  &process->m_drs,
+                  &dummy); /* unused; always 0 */
+   fclose(statmfile);
+
+   if (r == 7) {
+      process->super.m_virt *= host->pageSizeKB;
+      process->super.m_resident *= host->pageSizeKB;
+   }
+
+   return r == 7;
+}
```

Comment:
Have a look at [EnumProcessModulesEx](https://learn.microsoft.com/en-us/windows/win32/api/psapi/nf-psapi-enumprocessmodulesex).

Reply:
> Have a look at [EnumProcessModulesEx](https://learn.microsoft.com/en-us/windows/win32/api/psapi/nf-psapi-enumprocessmodulesex).

readMaps is only used to detect deleted libs, m_lrs is directly read from statm, not calculated.

for statm, cygwin implements it using kernel api directly: https://github.com/cygwin/cygwin/blob/a9e8e3d1cb8235f513f4d8434509acf287494fcf/winsup/cygwin/fhandler/process.cc#L1507

Reply:
I'd prefer for the htop port to take that route too (without the detour over cygwin's procfs layer).

Reply:
Maybe a quick note re `m_lrs` on Linux for reference:
On linux this field has not been reported for quite a while, thus `htop` implemented calculating it on-the-fly when this column is shown.

Libraries on Windows can't be really deleted/replaced while a process is using them, thus my remark re the two `mergedCmdline` flags.

The overall code thus should reduce quite a bit if using native API here directly.

Reply:
> Libraries on Windows can't be really deleted/replaced while a process is using them

i have tested it with a program which depends on a dll in cygwin, i am able to delete both the program executable and the library it loaded, and when the dll is deleted, its path showed in /proc/[pid]/maps starts with `C:\$Recycle.Bin` and some unique id in my case. in fact the deleted executable is also located there

Reply:
That's not really deleting the file, but moving (which Windows allows).

Reply:
> That's not really deleting the file, but moving (which Windows allows).

ohh, then should i remove the related code?